src/compiler/program.ts

```diff
@@ -857,11 +860,31 @@ namespace ts {
                 return oldProgram.structureIsReused;
             }
 
+            // If a file has ceased to be missing, then we need to discard some of the old
+            // structure in order to pick it up.
+            // Caution: if the file has created and then deleted between since it was discovered to
+            // be missing, then the corresponding file watcher will have been closed and no new one
+            // will be created until we encounter a change that prevents complete structure reuse.
+            // During this interval, creation of the file will go unnoticed.  We expect this to be
+            // both rare and low-impact.
+            if (oldProgram.getMissingFilePaths) {
+                const missingFilePaths: Path[] = oldProgram.getMissingFilePaths() || emptyArray;
+                for (const missingFilePath of missingFilePaths) {
+                    if (host.fileExists(missingFilePath)) {
+                        return oldProgram.structureIsReused = StructureIsReused.SafeModules;
+                    }
+                }
+            }
+
             // update fileName -> file mapping
             for (let i = 0; i < newSourceFiles.length; i++) {
                 filesByName.set(filePaths[i], newSourceFiles[i]);
             }
 
+            for (const p of oldProgram.getMissingFilePaths()) {

```

Comment:
I made the property optional in case there were existing implementers of the interface.  Was that over-cautious?